### PR TITLE
Document Collective SLIC Functions

### LINKED ISSUE #524: Clearly indicate which slic and lumberjack functions use collectives
The Slic and Lumbejack API docs should clearly indicate which slic functions/macros can use MPI collectives and which ones are always safe to call on specific ranks.

This could leak to Slic's `LogStream` derived types, such as `LumberjackStream`.

In particular, `slic::flush()` is a collective operation.

1. Make comment in the doxygen
2. Make a table in sphinx of calls which are collective, making it clear to users which ones have to called on all ranks

### PR DESCRIPTION
**Pull request by bmhan12**
This PR updates ReadTheDocs and doxygen to document which SLIC macros/functions are collective operations.
Resolves #524, relates to #26 discussion.

RTD of this branch:
https://axom.readthedocs.io/en/docs-han12-slicmpi/index.html

Table of collective SLIC macros:
https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#collective-slic-macros

Slic Doxygen link:
https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/namespaceaxom_1_1slic.html

Slic Macros Doxygen Link:
https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/slic__macros_8hpp.html

### CODE REVIEW

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:155]:
```suggestion
| **<RANK>**          | The MPI rank that emitted the message. Only applicable|
```
Please check and fix "emmit" elsewhere as well.

File: src/axom/slic/core/Logger.hpp
Comment (rhornung67) [inline src/axom/slic/core/Logger.hpp:176]: @bmhan12 since the comments you added to each of these methods is essentially the same, I think it would be more effective/impactful if the comment appeared once in a "prologue" and state that it applies to each of these methods. Specifically, use method grouping syntax in Doxygen like we do in SIdre, for example: https://github.com/LLNL/axom/blob/develop/src/axom/sidre/core/Group.hpp#L129

Also, I would move the part about the methods being collective operations and that all ranks must call them to the front before the qualifications based on message level.

File: src/axom/slic/docs/sphinx/sections/getting_started.rst
Comment (rhornung67) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:60]:
```suggestion
   If you do not initialize Slic, Slic will call ``slic::initialize()``,
```

File: src/axom/slic/docs/sphinx/sections/getting_started.rst
Comment (rhornung67) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:125]:
```suggestion
   No messages will be logged if you do not call ``slic::setLoggingMsgLevel()``.
```

File: src/axom/slic/interface/slic.hpp
Comment (rhornung67) [inline src/axom/slic/interface/slic.hpp:171]: Same comments that I made earlier about grouping the methods with Doxygen syntax and preface them with a single comment about their collective nature.

File: src/axom/slic/interface/slic_macros.hpp
Comment (rhornung67) [inline src/axom/slic/interface/slic_macros.hpp:24]: Similar comment about grouping methods in Doxygen and using single comment that applies to all.

**Review by rhornung67 — CHANGES_REQUESTED** [review]
@bmhan12 I made several comments suggesting what I think is a better way to group and document the methods. Most users tend to not read the details of lengthy comments on individual methods. If the main point is stated once and is clear that it applies to all grouped methods, it is easier for users to understand I think

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (rhornung67) [reply src/axom/slic/docs/sphinx/sections/architecture.rst:155]: Could be referring to Emmitt Smith? 😄

File: src/axom/slic/core/Logger.hpp
Comment (rhornung67) [inline src/axom/slic/core/Logger.hpp:169]: 👍

**Review by rhornung67 — APPROVED** [review]
Looks good @bmhan12 Thanks!  Others should review before merging to ensure all usage issues are addressed.

**Comment by agcapps** [comment]
I'll finish my review this morning.

File: src/axom/slic/core/Logger.hpp
Comment (agcapps) [inline src/axom/slic/core/Logger.hpp:175]: I like this grouping!  Thank you.  Rephrasing suggestion:
```suggestion
  //! The logMessage method is collective if:
```

File: src/axom/slic/docs/sphinx/sections/appendix.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:60]:
```suggestion
MPI-aware :ref:`LogStream` instances such as :ref:`SynchronizedStream`
```

File: src/axom/slic/docs/sphinx/sections/appendix.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:70]:
```suggestion
.. list-table:: Macros, Collective and Non-Collective
```

Additionally, throughout the table I'd change the word "when" to "after" (or perhaps "if," but I like that one less), although that's more of a style change to add specificity.

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:148]: THANK YOU

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:157]:
```suggestion
|                     | and with MPI-aware :ref:`LogStream` instances, such as,|
```
The class is `LogStream`, with a capital L, not `logStream`.  There may be other instances of this error throughout the SLIC docs.

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:293]:
```suggestion
The :ref:`LumberjackStream` is intended to be used with parallel MPI
```

File: src/axom/slic/docs/sphinx/sections/getting_started.rst
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/getting_started.rst:196]:
```suggestion
   MPI-aware :ref:`LogStream` instances such as :ref:`SynchronizedStream`
```
There may be others I didn't catch.

File: src/axom/slic/interface/slic.hpp
Comment (agcapps) [inline src/axom/slic/interface/slic.hpp:170]: See my comment on Logger.hpp at (new) line 175.

File: src/axom/slic/interface/slic_macros.hpp
Comment (agcapps) [inline src/axom/slic/interface/slic_macros.hpp:81]: This paragraph does need to be here: thank you for adding it.  Could you rephrase it? Currently, the paragraph starts out "These are collective macros."  Later it says "if abort on warning is enabled" and finished up with "default is disabled."

File: src/axom/slic/interface/slic_macros.hpp
Comment (agcapps) [inline src/axom/slic/interface/slic_macros.hpp:201]: See my comment on line 77.

**Review by agcapps — APPROVED** [review]
Thank you for this PR, @bmhan12 .  I gave a few suggestions, several of which are just style comments.

File: src/axom/slic/docs/sphinx/sections/appendix.rst
Comment (white238) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:90]: I would remove all the `(see Note)` since there are notes for all of the yes's and its right there.  Though I could be convinced to add them to all of them but my preference is to just remove it.

File: src/axom/slic/interface/slic_macros.hpp
Comment (white238) [inline src/axom/slic/interface/slic_macros.hpp:147]: This macro is not showing up properly in the doxygen:

https://axom.readthedocs.io/en/docs-han12-slicmpi/doxygen/html/slic__macros_8hpp.html#ad99a3eb01f4e3abe97a4ce625095f64e

Along with others.  It looks like the doxygen build needs `AXOM_DEBUG` enabled.

**Comment by white238** [comment]
Can you put something in the part where each macro is described that says this is collective in the appendix sphinx page:

https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#slic-error

**Comment by white238** [comment]
Can you remove all the links to "Axom toolkit" and just say Axom in the warnings like this one:

https://axom.readthedocs.io/en/docs-han12-slicmpi/axom/slic/docs/sphinx/sections/appendix.html#slic-assert

File: src/axom/slic/docs/sphinx/sections/appendix.rst
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/appendix.rst:74]: Nice table!

Since consecutive entries, e.g.  `SLIC_INFO` and `SLIC_INFO_IF` have the same values, would it make sense to combine these into a single row?

E.g. like this: https://stackoverflow.com/a/34222391

File: src/axom/slic/docs/sphinx/sections/architecture.rst
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/architecture.rst:94]: Suggestion (minor):
```suggestion
   All messages will be ignored until the first call to ``slic::setLoggingMsgLevel()``
```

Comment (kennyweiss) [reply src/axom/slic/docs/sphinx/sections/architecture.rst:94]: Thanks for calling this out.

Is this really the design though? 
Seems `slic::initialize()` should default to some logging for its first logger, e.g. like logging warnings and errors.

Perhaps we should create an issue to discuss this?

File: src/axom/slic/docs/sphinx/sections/wrapping_slic_in_macros.rst
Comment (kennyweiss) [inline src/axom/slic/docs/sphinx/sections/wrapping_slic_in_macros.rst:75]: 👍 

Minor: Should this be a `note` rather than a `warning`?

File: src/axom/slic/interface/slic_macros.hpp
Comment (kennyweiss) [reply src/axom/slic/interface/slic_macros.hpp:31]: 👍  for grouping.

I think it would also be helpful if the doxygen for the collective functions were tagged, e.g. with `\collective`, similar to how we use the `\accelerated` tags. This might make it easier for users to notice this when they are scrolling the header files rather than the generated doxygen.

**Review by kennyweiss — APPROVED** [review]
Thanks @bmhan12 -- this is a huge improvement!

Can you please go through the implementation of the functions to ensure that we're not missing any collective calls?
E.g. `slic::finalize()` ( and `Logger::finalize()`) is collective because it calls `flushStreams()` within it's implementation.
See: https://github.com/LLNL/axom/blob/608339cf0b9c48af3d302c4dcd8e947d475379e6/src/axom/slic/core/Logger.cpp#L333-L348

Also, I think it would be helpful if the doxygen for each function also included a `\collective` tag. Similarly to how the `\accelerated` tag was added in https://github.com/LLNL/axom/pull/620
This could be done in a follow-up PR (or not at all).

**Review by white238 — APPROVED** [review]

**Comment by white238** [comment]
Thanks @bmhan12 looks great!

**Pull request merged**
